<h1 style="background-color:orange; text-align:center; font-family:TimesNewRoman; font-weight:bold">UK ACCIDENT REPORT DATA ANALYSIS</h1>

<h2 style="text-align:center; font-family:TimesNewRoman; font-weight:bold;">ANALYST: CHRISTIAN R. NOBLEJAS</h2>

<h1 style="text-align:center; font-family:TimesNewRoman;">IMPORTING LIBRARIES NEEDED</h1>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import f_oneway

<h1 style="text-align:center; font-family:TimesNewRoman;">IMPORT DATASETS</h1>

In [2]:
accident = pd.read_csv("accident_data.csv")

In [3]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [4]:
accident.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

In [5]:
accident.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

<h1 style="text-align:center; font-family:TimesNewRoman; font-weight:bold;">NULL VALUES</h1>

In [6]:
accident['Latitude'] = accident['Latitude'].fillna(accident['Latitude'].mode()[0])
accident['Longitude'] = accident['Longitude'].fillna(accident['Longitude'].mode()[0])
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna('unknown road condition')
accident['Road_Type'] = accident['Road_Type'].fillna('unaccounted')
accident['Weather_Conditions'] = accident['Weather_Conditions'].fillna('unaccounted')
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].fillna(accident['Urban_or_Rural_Area'].mode()[0])
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

<h1 style="text-align:center; font-family:TimesNewRoman; font-weight:bold;">CONVERT DATA TYPE OF COLUMNS</h1>

In [7]:
accident['Index'] = accident['Index'].astype('category')
accident['Accident_Severity'] = accident['Accident_Severity'].astype('category')
accident['Light_Conditions'] = accident['Light_Conditions'].astype('category') 
accident['District Area'] = accident['District Area'].astype('category')
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].astype('category')
accident['Road_Type'] = accident['Road_Type'].astype('category')
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].astype('category')
accident['Weather_Conditions'] = accident['Weather_Conditions'].astype('category')
accident['Vehicle_Type'] = accident['Vehicle_Type'].astype('category')

In [8]:
accident.dtypes

Index                      category
Accident_Severity          category
Accident Date                object
Latitude                    float64
Light_Conditions           category
District Area              category
Longitude                   float64
Number_of_Casualties          int64
Number_of_Vehicles            int64
Road_Surface_Conditions    category
Road_Type                  category
Urban_or_Rural_Area        category
Weather_Conditions         category
Vehicle_Type               category
dtype: object

In [9]:
accident['Accident Date'] = accident['Accident Date'].astype('str')
accident['Accident Date'] = accident['Accident Date'].str.strip()
accident['Accident Date'] = accident['Accident Date'].str.replace('/','-')
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [10]:
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'],dayfirst =True, errors = 'coerce')
accident.dtypes

Index                            category
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                 category
District Area                    category
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions          category
Road_Type                        category
Urban_or_Rural_Area              category
Weather_Conditions               category
Vehicle_Type                     category
dtype: object

<h1 style="text-align:center; font-family:TimesNewRoman; font-weight:bold;">DETERMINING NULL VALUES AND FILL IT</h1>

In [11]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [12]:
accident['Year']=accident['Accident Date'].dt.year
accident['Month']=accident['Accident Date'].dt.month
accident['Day']=accident['Accident Date'].dt.day
accident['DayofWeek']=accident['Accident Date'].dt.dayofweek
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious    2019-06-05  51.506187   
1       200701BS65737           Serious    2019-07-02  51.495029   
2       200701BS66127           Serious    2019-08-26  51.517715   
3       200701BS66128           Serious    2019-08-16  51.495478   
4       200701BS66837            Slight    2019-09-03  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    2022-02-18  57.374005   
660675  201091NM01881            Slight    2022-02-21  57.232273   
660676  201091NM01935            Slight    2022-02-23  57.585044   
660677  201091NM01964           Serious    2022-02-23  57.214898   
660678  201091NM02142           Serious    2022-02-28  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2              unaccounted               Urban            unaccounted   
3       Single carriageway               Urban     Fine no high winds   
4              unaccounted               Urban            unaccounted   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  Year  Month  Day  DayofWeek  
0                                        Car  2019      6    5          2  
1                                        Car  2019      7    2          1  
2                      Taxi/Private hire car  2019      8   26          0  
3       Bus or coach (17 or more pass seats)  2019      8   16          4  
4                              Other vehicle  2019      9    3          1  
...                                      ...   ...    ...  ...  

In [13]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
Year                       0
Month                      0
Day                        0
DayofWeek                  0
dtype: int64

<h1 style="text-align:center; font-family:TimesNewRoman; font-size:50px">INSIGHTS</h1>

<h1>Q1: What is the most common accident severity category</h1>
<h2>Insight 1: The accident severity that has a highest record is Slight which has (563801) or 85.3% of 660679.</h2>

In [14]:
ac_sev = accident['Accident_Severity'].value_counts()
ac_sev

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64



<h2>Q2:What vehicle type is most frequently involved in accidents which result in Fatal?</h2>
<h2>Insight 2: The most common vehicle that was involved in an accidents is Car which has result of 6577 fatal accidents</h2>

In [15]:
vehicle_types = accident['Vehicle_Type'].groupby(accident['Accident_Severity']).value_counts().unstack()
vehicle_types.head(10)

C:\Users\DLL-AccountOfficer\AppData\Local\Temp\ipykernel_15280\3819518280.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  vehicle_types = accident['Vehicle_Type'].groupby(accident['Accident_Severity']).value_counts().unstack()


Vehicle_Type       Agricultural vehicle  Bus or coach (17 or more pass seats)  \
Accident_Severity                                                               
Fatal                                21                                   325   
Serious                             282                                  3373   
Slight                             1644                                 22180   

Vehicle_Type          Car  Data missing or out of range  \
Accident_Severity                                         
Fatal                6577                             0   
Serious             66461                             0   
Slight             424954                             6   

Vehicle_Type       Goods 7.5 tonnes mgw and over  \
Accident_Severity                                  
Fatal                                        216   
Serious                                     2321   
Slight                                     14770   

Vehicle_Type       Goods over 3.5t. and under 7.5t  \
Accident_Severity                                    
Fatal                                           67   
Serious                                        857   
Slight                                        5172   

Vehicle_Type       Minibus (8 - 16 passenger seats)  \
Accident_Severity                                     
Fatal                                            29   
Serious                                         276   
Slight                                         1671   

Vehicle_Type       Motorcycle 125cc and under  Motorcycle 50cc and under  \
Accident_Severity                                                          
Fatal                                     189                         95   
Serious                                  2031                       1014   
Slight                                  13049                       6494   

Vehicle_Type       Motorcycle over 125cc and up to 500cc  \
Accident_Severity                                          
Fatal                                                105   
Serious                                             1014   
Slight                                              6537   

Vehicle_Type       Motorcycle over 500cc  Other vehicle  Pedal cycle  \
Accident_Severity                                                      
Fatal                                339             70            6   
Serious                             3457            767           39   
Slight                             21861           4800          152   

Vehicle_Type       Ridden horse  Taxi/Private hire car  \
Accident_Severity                                        
Fatal                         0                    155   
Serious                       0                   1771   
Slight                        4                  11368   

Vehicle_Type       Van / Goods 3.5 tonnes mgw or under  
Accident_Severity                                       
Fatal                                              467  
Serious                                           4554  
Slight                                           29139

<h2>Q3:On which road type do most car accidents occur?</h2>
<h2>Insight 3: Based on the result, Mostly car accidents occur on a single carriageways</h2>

In [16]:
car_acc = accident[accident['Vehicle_Type'] == 'Car']
road_co = car_acc['Road_Type'].value_counts()
road_co

Road_Type
Single carriageway    371028
Dual carriageway       74820
Roundabout             33187
One way street         10223
Slip road               5298
unaccounted             3436
Name: count, dtype: int64

<h2>Q4:On which road_surface do most car accidents occur?</h2>
<h2>Insight 4: Most car accidents occur on dry roads, It's either a human error or due to traffic load .</h2>


In [17]:
road_surface= car_accidents['Road_Surface_Conditions'].value_counts()
road_surface


NameError: name 'car_accidents' is not defined



<h2>Q5: Which year got the highest record of Accidents?</h2>
<h2>Insight 5: On year 2019 got the Higher number or count of Accidents</h2>

In [ ]:
car_yr = car_accidents['Year'].value_counts()
car_yr


<h2>Q6: Does weather conditions affect a accidents?</h2>
<h2>Insight 6: Most car accidents occur in fine weather with no high winds, suggesting that human error.</h2>

In [ ]:
weather = car_accidents['Weather_Conditions'].value_counts()
weather


<h2>Q7: How many fatal accident records occurred under fog or mist conditions?</h2>
<h2>Insight 7: The total number of Fatal accidents occured during fog or mist conditions is 82 which is 12.1% of 677 total accidents.</h2>

In [ ]:
we_acc = accident[(accident['Weather_Conditions']== "Fog or mist") & (accident['Accident_Severity']=='Fatal')]
we_acc.shape[0]


<h2>Q8: How many fatal accident records occurred under fog or mist conditions?</h2>
<h2>Insight 8: The total number of Fatal accidents occurred during Fine no high winds conditions is 7100 which is 1.8% of total accident of 392668.</h2>

In [ ]:
we_acc = accident[(accident['Weather_Conditions']== "Fine no high winds") & (accident['Accident_Severity']=='Fatal')]
we_acc.shape[0]

<h2> Q9: Which Year got the highest record of Fatal accidents?</h2>

<h2>Insight 9: In 2019, it had the highest record of Fatal accidents</h2>




In [ ]:
fatal_yr = accident[accident['Accident_Severity'] == 'Fatal'].groupby('Year').size()
fatal_yr

<h2> Q10: What are the top 5 District Areas that have the highest records of accidents?</h2>
<h2>Insight 10: This area got the highest record of accidents, these are Birmingham, Leeds, Manchester, Bradford, and Sheffield</h2>




In [ ]:
dis_area = accident['District Area'].value_counts().head(5).index

<h2>Q11: Which District area on the top 5 has the highest fatality rate record?</h2>
<h2>Insight 11: The District with the highest record of Fatal accidents is Birmingham has 105 fatal accidents, while Sheffield has 38 fatal accidents.  </h2>

In [ ]:
fatal_counts = accident[accident["Accident_Severity"] == "Fatal"]['District Area'].value_counts()
fatal_counts[dis_area]


<h2>Q12: On the top 5 district areas, do the weather conditions affect the number of Fatal accidents?</h2>
<h2>Insight 11: Based on the results, Birmingham recorded 82 fatal accidents, accounting for 82.1% of the total 105 accidents in the area. </h2>

In [ ]:
fatal_accidents = accident[accident["Accident_Severity"] == "Fatal"]
fatal_counts_by_weather = fatal_accidents.groupby(["District Area", "Weather_Conditions"]).size().unstack()
top5_areas = accident['District Area'].value_counts().head(5).index  
fatal_counts_by_weather.loc[top5_areas]


<h2> Q13: On which area (Urban or Rural) had the highest fatality rate?</h2>

<h2>Rural areas have more fatal accidents than urban areas, even though urban areas have more total number of vehicles. </h2>

In [ ]:
ur_sev = accident[accident['Accident_Severity'] == 'Fatal']['Urban_or_Rural_Area'].value_counts()
ur_sev



<h2>Q14: Do light conditions affect the severity of an accident?</h2>
<h2>Insight 14: Most fatal accidents happen in daylight due to reasons like traffic, but many also occur in total darkness without lighting. Serious and slight accidents are more common in  dark lighting.</h2>

In [ ]:
sev_light_casual = accident.groupby(['Accident_Severity', 'Light_Conditions'])['Number_of_Casualties'].sum().unstack()
sev_light_casual


<h2>Q15: Do road surface conditions impact accidents in urban and rural areas?</h2>

<h3>Insight 15: Most accidents happen on dry roads, but wet conditions also cause many accidents. Rural areas have more accidents on icy roads than urban areas, likely due to less road maintenance.</h3>

In [ ]:
ruban_sur = accident.groupby('Urban_or_Rural_Area')['Road_Surface_Conditions'].value_counts().unstack()
ruban_sur

<h2>Q16: What month had the highest number of accidents?</h2>

<h2>Insight 16: In November had the highest number of accidents due to heavy traffic and christmas preparation.</h2>

In [ ]:
mon_acc = accident['Month'].value_counts()
mon_acc 

<h2>Insight 17: The total number of Casualties each year, In 2019 had the highest record og casualties</h2>

In [ ]:
tot_cas = accident.groupby(['Year'])['Number_of_Casualties'].sum()
tot_cas 

<h2> Q18: Which month has have highest accident rate?</h2>
<h2>Insight 18: November had the highest number of accidents in 2019, with 16,559 accidents, accounting for 1.38% of the total 1,198,591 casualties.</h2>


In [ ]:
yr_2019 = accident[accident['Year'] == 2019]


In [ ]:
yr_2019['Month'].sum()

In [ ]:
yr_2019['Month'].value_counts()

<h2>Insight 19: On year 2019, Dry road was the most dangerous road surface as it gained 125,348 accidents</h2>



In [ ]:
we_con = yr_2019['Road_Surface_Conditions'].value_counts()
we_con

<h2>Insight 20: In 2019, Accidents are more likely to happen in Dry road areas. Let's assume that it was human error, vehicle malfunction, or Distracted while driving.
</h2>



In [ ]:
road_sev= yr_2019.groupby('Road_Surface_Conditions')['Accident_Severity'].value_counts()
road_sev